File: C/2015/day02.c
#include "readFileToString.h"
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define DATA_FILE "day02.dat"

enum dimentions {
	LENGTH = 0, WIDTH, HEIGHT
};

int
day02()
{
	const char     *filename = DATA_FILE;
	char	       *fileContents = readFileToString(filename);

	if (fileContents) {
		//printf("%s\n", fileContents);

		int		total_area = 0;
		int		total_ribbon = 0;

		const char	outer_delimiters[] = "\n";
		const char	inner_delimiters[] = "x";

		char	       *token;
		char	       *outer_saveptr = NULL;
		char	       *inner_saveptr = NULL;

		token = strtok_r(fileContents, outer_delimiters, &outer_saveptr);

		while (token != NULL) {
			//printf("Outer Token: %s\n", token);

			char	       *inner_token =
			strtok_r(token, inner_delimiters, &inner_saveptr);

			int		i = 0;
			int		dim[] = {LENGTH, WIDTH, HEIGHT};

			while (inner_token != NULL) {
				//printf("Inner Token: %s\n", inner_token);
				dim[i++] = atoi(inner_token);
				inner_token = strtok_r(NULL, inner_delimiters, &inner_saveptr);
			}

			int		smallest = dim[LENGTH] * dim[WIDTH];
			total_area += 2 * smallest;
			int		width_height = dim[WIDTH] * dim[HEIGHT];
			smallest = smallest < width_height ? smallest : width_height;
			total_area += 2 * width_height;
			int		height_length = dim[HEIGHT] * dim[LENGTH];
			smallest = smallest < height_length ? smallest : height_length;
			total_area += 2 * height_length;
			total_area += smallest;

			smallest = 2 * dim[LENGTH] + 2 * dim[WIDTH];
			width_height = 2 * dim[WIDTH] + 2 * dim[HEIGHT];
			smallest = smallest < width_height ? smallest : width_height;
			height_length = 2 * dim[HEIGHT] + 2 * dim[LENGTH];
			smallest = smallest < height_length ? smallest : height_length;
			total_ribbon += dim[LENGTH] * dim[WIDTH] * dim[HEIGHT];
			total_ribbon += smallest;

			token = strtok_r(NULL, outer_delimiters, &outer_saveptr);
		}

		printf("2015 Day 02 Part 1: %d\n", total_area);
		printf("2015 Day 02 Part 2: %d\n", total_ribbon);

		free(fileContents);
	}

	return 0;
}

File: zig/2015/day02.zig
const std = @import("std");

pub fn main() !void {
    // 1. Get a handle to the current working directory.
    const cwd = std.fs.cwd();

    // 2. Open the file in the current working directory.
    const file = try cwd.openFile("../../data/2015/day02.dat", .{});
    defer file.close();

    // 3. Create a GeneralPurposeAllocator instance.
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer {
        const leak_check = gpa.deinit();
        if (leak_check == .leak) {
            std.debug.print("Memory leak detected!\n", .{});
        }
    }
    const allocator = gpa.allocator();

    // 4. Read the entire file into an allocated buffer (new Zig 0.15.x API).
    const file_contents = try file.readToEndAlloc(allocator, std.math.maxInt(usize));
    defer allocator.free(file_contents);

    // 5. Print the contents.
    // std.debug.print("File contents:\n{s}\n", .{file_contents});

    var total_area: u32 = 0;
    var total_ribbon: u32 = 0;
    var it = std.mem.splitScalar(u8, file_contents, '\n');

    while (it.next()) |line| {
        if (line.len == 0) continue; // skip empty lines

        var dims = std.mem.splitScalar(u8, line, 'x');

        const length_str = dims.next().?;
        const width_str = dims.next().?;
        const height_str = dims.next().?;

        const length = try std.fmt.parseInt(u32, length_str, 10);
        const width = try std.fmt.parseInt(u32, width_str, 10);
        const height = try std.fmt.parseInt(u32, height_str, 10);

        var smallest = length * width;
        total_area += 2 * smallest;
        var width_height = width * height;
        smallest = if (smallest < width_height) smallest else width_height;
        total_area += 2 * width_height;
        var height_length = height * length;
        smallest = if (smallest < height_length) smallest else height_length;
        total_area += 2 * height_length;
        total_area += smallest;

        smallest = 2 * length + 2 * width;
        width_height = 2 * width + 2 * height;
        smallest = if (smallest < width_height) smallest else width_height;
        height_length = 2 * height + 2 * length;
        smallest = if (smallest < height_length) smallest else height_length;
        total_ribbon += length * width * height;
        total_ribbon += smallest;
    }

    std.debug.print("Total area: {d}\n", .{total_area});
    std.debug.print("Total ribbon: {d}\n", .{total_ribbon});
}
